pallets/nft-marketplace/src/lib.rs

```
@@ -0,0 +1,330 @@
+		pub fn remove(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			let sale = <Gallery<T>>::get(class_id, instance_id).ok_or(Error::<T>::NotForSale)?;
+
+			// Ensure that the buyer is not the seller of the NFT
+			ensure!(who == sale.seller, Error::<T>::NotOwner);
+
+			// Transfer the NFT back to the seller, i.e., the original owner of this NFT
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &sale.seller)
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
+
+			// Remove the NFT from the gallery
+			<Gallery<T>>::remove(class_id, instance_id);
+			Self::deposit_event(Event::Removed {
+				class_id,
+				instance_id,
+			});
+
+			Ok(())
+		}
+
+		/// Buy the given nft
+		///
+		/// Fails if
+		///   - the NFT is not for sale
+		///   - `origin` is the seller of the NFT
+		///   - `origin` does not have enough balance of the currency the nft is being sold in
+		///   - transferring the asking price from the buyer to the seller fails
+		///   - transferring the nft to the buyer fails
+		#[pallet::weight(10_000_000)]
+		#[transactional]
+		pub fn buy(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		) -> DispatchResult {
+			let buyer = ensure_signed(origin.clone())?;
+			let sale = <Gallery<T>>::get(class_id, instance_id).ok_or(Error::<T>::NotForSale)?;
+
+			// Ensure that the buyer is not the seller of the NFT
+			ensure!(buyer != sale.seller, Error::<T>::IsSeller);
+
+			// Make sure the buyer can pay for the NFT
+			T::Fungibles::can_withdraw(sale.price.currency, &buyer, sale.price.amount)
+				.into_result()
+				.map_err(|_| Error::<T>::InsufficientBalance)?;
+
+			// Have the buyer pay the seller for the NFT
+			T::Fungibles::transfer(
+				sale.price.currency,
+				&buyer,
+				&sale.seller,
+				sale.price.amount,
+				true,
+			)
+			.map_err(|_| Error::<T>::PaymentFailed)?;
```

We probably don't want to `map_err` here - it's better to expose the full detailed error of the underlying currency pallet, IMO

---

Thanks for pointing this out! I oversaw how errors worked in here. I assumed the `Error` type of `DispatchResult` would be our own `Error` type and thus we needed to return one of our variants. I tried to use `thiserror` to wrap the underlying error in all of these variants but Substrate doesn't allow for that. But I now understand that we can just return the underlying error directly which is definitely much better :)